src/module.cpp

```diff
@@ -272,10 +272,23 @@ int Module::CompileFile() {
             fclose(f);
         }
 
-        std::string buffer;
-        llvm::raw_string_ostream os(buffer);
-        execPreprocessor(!IsStdin(filename) ? filename : "-", &os);
-        YY_BUFFER_STATE strbuf = yy_scan_string(os.str().c_str());
+        // If the CPP stream has been initialized, we have unexpected behavior.
+        if (bufferCPP) {
+            perror(filename);
+            return 1;
+        }
+
+        // Replace the CPP stream with a newly allocated one.
+        bufferCPP.reset(new CPPBuffer{});
+
+        const int numErrors = execPreprocessor(!IsStdin(filename) ? filename : "-", bufferCPP->os.get());
+        errorCount += (g->ignoreCPPErrors) ? 0 : numErrors;
+
+        if (g->onlyCPP) {
+            return errorCount; // Return early
+        }
+
+        YY_BUFFER_STATE strbuf = yy_scan_string(bufferCPP->str.c_str());
```

Comment:
Old behavior - `std::string buffer` is destroyed when the scope is closed. New behavior - `bufferCPP` still lives will the `Module` is destroyed. Am I missing something?

Reply:
See clean up [here](https://github.com/ispc/ispc/pull/2280/files#r833824859). I think this is sufficient to release the memory, but I may have misaccounted.

Basically, the issue is that we need to keep the string buffer around until the call to `writeOutput`/`writeCPPStub`. I allocated a new pointer for every call to `execPreprocessor`, which should be accompanied by exactly one call to `reset` in `writeCPPStub` once safe to do so.

Reply:
See commit 006bf80.

Reply:
This doesn't seem to solve the issue.

What seems to be ok'ish solution - not very elegant, but good enough, is the following:
- buffer is pointer to a wrapper class in the Module (as it is now) and it is initialized as nullptr
- create buffer before calling the preprocessor (as it is now)
- if we go regular compilation - destroy buffer after the parser (after yy_delete_buffer())
- if it's -E compilation - destroy the buffer after it's used in WriteCPPStub.
- assert in WriteCPPStub that buffer valid nullptr.

Reply:
> * if it's -E compilation - destroy the buffer after it's used in WriteCPPStub.
> * assert in WriteCPPStub that buffer valid nullptr.

See [comment](https://github.com/ispc/ispc/pull/2280/files#r834598949).